invoices/update_invoice.go

```diff
@@ -0,0 +1,778 @@
+			return nil, err
+		}
+
+		return updateCtx.Invoice, nil
+
+	case SettleHodlInvoiceUpdate:
+		updateCtx, err := settleHodlInvoice(
+			invoice, hash, updateTime, update.State,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		err = updater.StoreSettleHodlInvoiceUpdate(*updateCtx)
+		if err != nil {
+			return nil, err
+		}
+
+		return updateCtx.Invoice, nil
+
+	case CancelInvoiceUpdate:
+		updateCtx, err := cancelInvoice(
+			invoice, hash, updateTime, update.State,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		err = updater.StoreCancelInvoiceUpdate(*updateCtx)
+		if err != nil {
+			return nil, err
+		}
+
+		return updateCtx.Invoice, nil
+
+	default:
+		return nil, fmt.Errorf("unknown update type: %s",
+			update.UpdateType)
+	}
+}
+
+// cancelHTLCs tries to cancel the htlcs in the given InvoiceUpdateDesc.
+//
+// NOTE: cancelHTLCs updates will only use the `CancelHtlcs` field in the
+// InvoiceUpdateDesc.
+func cancelHTLCs(invoice *Invoice, updateTime time.Time,
+	update *InvoiceUpdateDesc) (*InvoiceUpdaterContext, error) {
+
+	// Process add actions from update descriptor.
+	htlcsAmpUpdate := make(map[SetID]map[models.CircuitKey]*InvoiceHTLC)
+
+	// Process cancel actions from update descriptor.
+	cancelHtlcs := update.CancelHtlcs
+	for key, htlc := range invoice.Htlcs {
+		htlc := htlc
+
+		// Check whether this htlc needs to be canceled. If it does,
+		// update the htlc state to Canceled.
+		_, cancel := cancelHtlcs[key]
+		if !cancel {
+			continue
+		}
+
+		err := cancelSingleHtlc(updateTime, htlc, invoice.State)
+		if err != nil {
+			return nil, err
+		}
+
+		// Delete processed cancel action, so that we can check later
+		// that there are no actions left.
+		delete(cancelHtlcs, key)
+
+		// Tally this into the set of HTLCs that need to be updated on
+		// disk, but once again, only if this is an AMP invoice.
+		if invoice.IsAMP() {
+			cancelHtlcsAmp(invoice, htlcsAmpUpdate, htlc, key)
+		}
+	}
+
+	// Verify that we didn't get an action for htlcs that are not present on
+	// the invoice.
+	if len(cancelHtlcs) > 0 {
+		return nil, errors.New("cancel action on non-existent htlc(s)")
+	}
+
+	return &InvoiceUpdaterContext{
+		Invoice:        invoice,
+		UpdateTime:     updateTime,
+		AMPHTLCsUpdate: htlcsAmpUpdate,
+	}, nil
+}
+
+// addHTLCs tries to add the htlcs in the given InvoiceUpdateDesc.
+func addHTLCs(invoice *Invoice, hash *lntypes.Hash,
+	updateTime time.Time, update *InvoiceUpdateDesc) (
+	*InvoiceUpdaterContext, error) {
+
+	var setID *[32]byte
+	invoiceIsAMP := invoice.IsAMP()
+	if invoiceIsAMP && update.State != nil {
+		setID = update.State.SetID
+	}
+
+	// Process add actions from update descriptor.
+	htlcsAmpUpdate := make(map[SetID]map[models.CircuitKey]*InvoiceHTLC) //nolint:lll
+	for key, htlcUpdate := range update.AddHtlcs {
+		if _, exists := invoice.Htlcs[key]; exists {
+			return nil, fmt.Errorf("duplicate add of htlc %v", key)
+		}
+
+		// Force caller to supply htlc without custom records in a
```

Comment:
Since this commit is just moving code around, perhaps best to recreate these comments on other commits as every time I rebase I have to manually recreate this commit. Also maybe just to reduce the scope of the PR, perhaps we could do these refactors in a follow up. WDYT?